### PR TITLE
Jianfeng/fix ssh script in startcc

### PR DESCRIPTION
**Pull request by JavierJia**
The real problem of getip.sh is that it return a `stdout: slave: localhost  23480` , that is really weird.
I add a default cmd to give the result of `hostname -i` if the stdout result is invalide.

It worked in this build : https://travis-ci.org/uci-cbcl/genomix/builds/13068366

### CODE REVIEW

**Comment by jakebiesinger** [comment]
stdout in this case should be the hostname from the master file, then tab, then PID of the CC.  We record the PID and use it to do a graceful shutdown later (that e.g., doesn't kill all the CC's on this machine).

I'll check it out and try it out soon.

File: genomix/genomix-driver/src/main/resources/scripts/getip.sh [hunk 1]
```diff
@@ -4,25 +4,27 @@
...
 #    then
-        IPADDR=`/sbin/ifconfig eth0 | grep "inet " | awk '{print $2}' | cut -f 2 -d ':'`
+        IPADDR=`ifconfig eth0 | grep "inet " | awk '{print $2}' | cut -f 2 -d ':'`
         if [ "$IPADDR" = "" ]
         then
-            IPADDR=`/sbin/ifconfig lo | grep "inet " | awk '{print $2}' | cut -f 2 -d ':'`
+            IPADDR=`ifconfig lo | grep "inet " | awk '{print $2}' | cut -f 2 -d ':'`
         fi 
+        IS_VALID_IP=`echo $IPADDR | grep -E "([0-9]{1,3}\.{0,1}){4}"`
+        [ "$IS_VALID_IP" ] || IPADDR=`hostname -i`
```
Comment (jakebiesinger) [inline genomix/genomix-driver/src/main/resources/scripts/getip.sh:20]: Can you do this check at the very end? The if/else is trying to be more platform-independent.  Why skip the `en1` business?

Comment (JavierJia) [inline genomix/genomix-driver/src/main/resources/scripts/getip.sh:20]: Actually I found that this build still failed. That 'eth0 not found' was failed in every build. Something the program still run and pass, sometime it didn't. 
The real reason of the failing I think it should not be so related to getip.sh. 

I've copied the Yingyi to see if he has some hint.  Do you have some idea ? 
The exception track is below:

```
Exception in thread "main" edu.uci.ics.hyracks.api.exceptions.HyracksException: edu.uci.ics.hyracks.api.exceptions.HyracksException: Job failed on account of:
Unsatisfiable number of partitions for PARTITION_COUNT(ODID:0): 0
    at edu.uci.ics.pregelix.core.driver.Driver.runJobs(Driver.java:160)
    at edu.uci.ics.pregelix.core.driver.Driver.runJobs(Driver.java:76)
    at edu.uci.ics.genomix.driver.GenomixDriver.flushPendingJobs(GenomixDriver.java:234)
    at edu.uci.ics.genomix.driver.GenomixDriver.runGenomix(GenomixDriver.java:289)
    at edu.uci.ics.genomix.driver.GenomixDriver.main(GenomixDriver.java:339)
Caused by: edu.uci.ics.hyracks.api.exceptions.HyracksException: Job failed on account of:
Unsatisfiable number of partitions for PARTITION_COUNT(ODID:0): 0
    at edu.uci.ics.hyracks.control.cc.job.JobRun.waitForCompletion(JobRun.java:193)
    at edu.uci.ics.hyracks.control.cc.work.WaitForJobCompletionWork$2.run(WaitForJobCompletionWork.java:58)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:722)
Caused by: edu.uci.ics.hyracks.api.exceptions.HyracksException: Unsatisfiable number of partitions for PARTITION_COUNT(ODID:0): 0
    at edu.uci.ics.hyracks.control.cc.scheduler.ActivityClusterPlanner.computePartitionCounts(ActivityClusterPlanner.java:406)
    at edu.uci.ics.hyracks.control.cc.scheduler.ActivityClusterPlanner.planActivityCluster(ActivityClusterPlanner.java:65)
    at edu.uci.ics.hyracks.control.cc.scheduler.JobScheduler.findRunnableTaskClusterRoots(JobScheduler.java:131)
    at edu.uci.ics.hyracks.control.cc.scheduler.JobScheduler.findRunnableTaskClusterRoots(JobScheduler.java:100)
    at edu.uci.ics.hyracks.control.cc.scheduler.JobScheduler.startRunnableActivityClusters(JobScheduler.java:156)
    at edu.uci.ics.hyracks.control.cc.scheduler.JobScheduler.startJob(JobScheduler.java:94)
    at edu.uci.ics.hyracks.control.cc.work.JobStartWork.doRun(JobStartWork.java:64)
    at edu.uci.ics.hyracks.control.common.work.SynchronizableWork.run(SynchronizableWork.java:32)
    at edu.uci.ics.hyracks.control.common.work.WorkQueue$WorkerThread.run(WorkQueue.java:116)
```

Comment (jakebiesinger) [inline genomix/genomix-driver/src/main/resources/scripts/getip.sh:20]: If the error is that there are no partitions, then the either the nc hasn't
started yet or it has failed (cat /tmp/t2/_/_.log). Perhaps the /temp
directory isn't writable or perhaps the memory requested is larger than is
available.

There's a command line wait time that will cause the driver to wait longer
before submitting work to the cc. If it's just a timing thing, that's where
I'd start. And see what is in those log files for the nc.

I wish we had a way to listen to those logs from the driver. And to know if
the nc's actually started correctly.

File: .travis.yml [hunk 2]
```diff
@@ -5,6 +5,8 @@ jdk:
     - openjdk7
 
 install:
+    - ifconfig
+    - ./genomix/genomix-driver/src/main/resources/scripts/getip.sh
```
Comment (jakebiesinger) [inline .travis.yml:9]: Is this script marked as executable on a fresh checkout? git should be able to handle making it so if it's not already.

File: genomix/genomix-driver/src/main/resources/scripts/startDebugNc.sh [hunk 3]
```diff
@@ -10,7 +10,7 @@ cd "$GENOMIX_HOME"
 
 #Get the IP address of the cc
 CCHOST_NAME=`cat conf/master`
-CCHOST=`ssh ${CCHOST_NAME} "cd ${GENOMIX_HOME}; bin/getip.sh"`
+CCHOST=`ssh ${CCHOST_NAME} 'bash -s' < ${GENOMIX_HOME}/bin/getip.sh`
```
Comment (jakebiesinger) [inline genomix/genomix-driver/src/main/resources/scripts/startDebugNc.sh:13]: Can you explain this change?  why not just

``` bash
CCHOST=`ssh -n ${CCHOST_NAME} "${GENOMIX_HOME}/bin/getip.sh"`
```

File: .travis.yml (hunk 2 above)
Comment (JavierJia) [inline .travis.yml:9]: Yep, you are correct. This is for the debug purpose to test if getip works on travis machine.

File: genomix/genomix-driver/src/main/resources/scripts/startDebugNc.sh (hunk 3 above)
Comment (JavierJia) [inline genomix/genomix-driver/src/main/resources/scripts/startDebugNc.sh:13]: The reason is that the "${GENOMIX_HOME}" could not get proper translated. (Always empty string).

**Comment by JavierJia** [comment]
@jakebiesinger , how to undo this pull request? 
The source of the Travis build failed still not clear. I don't want to merge...

File: genomix/genomix-driver/src/main/resources/scripts/startDebugNc.sh (hunk 3 above)
Comment (jakebiesinger) [inline genomix/genomix-driver/src/main/resources/scripts/startDebugNc.sh:13]: then there's something else wrong.  The string expansion should happen locally using the above-set value for `$GENOMIX_HOME`. `ssh` should only see the fully expanded variable, not $GENOMIX_HOME at all.  Did you have it in single-quotes or something?

**Comment by jakebiesinger** [comment]
There's no "undo" for a pull request but there is a "close" which means "won't merge".  With my incoming cluster changes, we should get a better idea-- if the NC doesn't connect, there will (now) be an error with a tail of the log file.

I'm closing this request. Perhaps you'd like to also delete the branch?

**Pull request closed without merging**